### PR TITLE
Add no worker timeout for scheduler

### LINKED ISSUE #8126: Eventually shut down scheduler when no workers return
Currently, a scheduler is not considered idle if it has tasks waiting to be processed but no workers that could process them: https://github.com/hendrikmakait/distributed/blob/4b10aa7068ff8a67cc756ef0c91f5553429963f3/distributed/scheduler.py#L7966-L7972

This makes sense in adaptive scaling scenarios or when using ephemeral workers to give new workers time to spin up. However, we should eventually shut down the scheduler if no workers return instead of leaving it running indefinitely.

We could do this by adjusting the definition of the idle timeout to include time spent without workers or adding a dedicated timeout mechanism.

### PR DESCRIPTION
**Pull request by FTang21**
Closes #8126

- [x] Tests added / passed
- [X] Passes `pre-commit run --all-files`

Add a new variable no_worker_timeout that checks if there are tasks still waiting to be processed, but no worker is processing them. 
Update logic to include this behavior without changing the old logic too much.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

File: distributed/scheduler.py
Comment (FTang21) [reply distributed/scheduler.py]: updated to self.no_worker_since = time() to remain consistent with self.idle_since, this will still be set to None when workers are processing tasks.

**Comment by hendrikmakait** [comment]
@fjetter: Do you have time for another review on this PR?

**Comment by fjetter** [comment]
Good looks fine but there are a couple of test failures that appear to be related

There is a test failure in `distributed.tests.test_config` that raises a config schema error. That's likely connected to the new option.

`test_idle_timeout_no_workers` also failed once. This is likely a little flaky but I'd be surprised if this isn't connected. I recommend running the test a couple of times concurrently. This typically triggers those flaky things reliably, e.g. `pytest  -x distributed/tests/test_scheduler.py::test_idle_timeout_no_workers -nauto --count=100` (requires `pytest-xdist` and `pytest-repeat`)

**Comment by crusaderky** [comment]
@FTang21 do you have time to look into the oustanding issues?

File: distributed/distributed-schema.yaml
Comment (crusaderky) [inline distributed/distributed-schema.yaml:89]: Could you please add support for null to disable it?

File: distributed/scheduler.py
Comment (crusaderky) [inline distributed/scheduler.py:8169]: This crashes if idle-timeout-no-worker is null

File: distributed/tests/test_scheduler.py
Comment (crusaderky) [inline distributed/tests/test_scheduler.py:2416]: Please add test cases:
- config is set to null
- there are actually no workers; tasks have no worker restrictions so they end up in queued instead of unrunnable

File: distributed/tests/test_scheduler.py
Comment (crusaderky) [inline distributed/tests/test_scheduler.py:2411]: This is quite fragile.

```suggestion
    await async_poll_for(s.check_idle, timeout=3)
```

**Review by crusaderky — CHANGES_REQUESTED** [review]

File: distributed/distributed.yaml
Comment (crusaderky) [inline distributed/distributed.yaml:21]: I think the default should be coherent with idle-timeout.

**Comment by crusaderky** [comment]
I can envision how users may get confused by the cluster shutting down unexpectedly. Could you please add logging that explain which timeout tripped the timeout?

File: distributed/scheduler.py
Comment (crusaderky) [inline distributed/scheduler.py:3866]: I would be annoyed if I had set a timeout of 2h and the shutdown was 30min later than expected.
These methods are tiny and running them 4 times per second is inconsequential.

File: distributed/scheduler.py
Comment (hendrikmakait) [inline distributed/scheduler.py:3524]: In the spirit of #8190, is there a reason to make these public?

File: distributed/tests/test_scheduler.py
Comment (hendrikmakait) [inline distributed/tests/test_scheduler.py:2446]: Did you mean
```suggestion
    config={"distributed.scheduler.no-workers-timeout": None},
```
?

File: distributed/scheduler.py
Comment (hendrikmakait) [inline distributed/scheduler.py:8197]: Should we also check whether we have tasks in processing (regardless of `queued` tasks) and return if that's the case?

Comment (crusaderky) [reply distributed/scheduler.py:8197]: I don't think so.
The intent is to shut down the cluster if there are tasks to run but nowhere to run them.

"there's a task to run" translates to `queued`,`unrunnable`, or `processing`.

- `unrunnable` tasks exist if there are no workers whatsoever and queueing is disabled, or if there are _some_ workers but they've all been excluded by task restrictions. In both cases their existance should trip the timeout.
- `queued` tasks exist either if there are no workers whatsoever or all workers are busy with other tasks. So if there are any workers on the cluster, we can just assume they're already busy with other rootish tasks and reset the timeout.
- `processing` tasks exist if there's somewhere to run them, which should cause the timeout to reset.

Comment (hendrikmakait) [reply distributed/scheduler.py:8197]: Doesn't that translate to:

```suggestion
        if self.processing or (not self.queued and not self.unrunnable) or (self.queued and self.workers):
```

Comment (hendrikmakait) [reply distributed/scheduler.py:8197]: Also, let me rephrase: 
Should we also check whether we have tasks in processing (regardless of queued tasks) and reset `no_workers_since` if that's the case?

File: distributed/scheduler.py
Comment (crusaderky) [reply distributed/scheduler.py:3524]: As a rule of thumb it makes sense to me to keep the variables that hold a public setting as public too.
I don't have a strong opinion on the two `_since` variables or the two `check_` methods, so I'll change them to private (without deprecation, if you agree that there shouldn't be anybody hooking into them).

Comment (crusaderky) [reply distributed/scheduler.py:3524]: I'll change the pre-existing attributes in a follow-up PR though.

**Comment by crusaderky** [comment]
@hendrikmakait all comments have been addressed

File: distributed/scheduler.py
Comment (crusaderky) [reply distributed/scheduler.py:8197]: `SchedulerState.processing` sadly doesn't exist; you'd need `any(ws.processing for ws in self.workers.values())` (like `check_idle` does).

Do you see a use case that the current logic doesn't cover? I can't think of any...

Comment (hendrikmakait) [reply distributed/scheduler.py:8197]:
```python
@gen_cluster(
    client=True,
    nthreads=[("", 1)],
    config={"distributed.scheduler.no-workers-timeout": "100ms"},
)
async def test_no_workers_timeout_with_worker(c, s, a):
    """Do not trip no-workers-timeout when there are tasks processing"""
    import time
    s._check_no_workers()
    await asyncio.sleep(0.2)
    assert s.status == Status.running

    f1 = c.submit(time.sleep, 2)
    f2 = c.submit(inc, 1, key="x", workers=["127.0.0.2:1234"])
    await f1
    assert s.status == Status.running
```

would kill the scheduler before we're able to complete `f1`. From what I understand, we only ever want to kill the cluster if there's nothing that could possibly be processed with the current set of workers.

**Review by hendrikmakait — CHANGES_REQUESTED** [review]

File: distributed/scheduler.py
Comment (crusaderky) [reply distributed/scheduler.py]: :eyes: I had not seen _that_ use case. Thank you. Fixed.

File: distributed/distributed.yaml
Comment (crusaderky) [reply distributed/distributed.yaml]: Well, there's the use case of adaptive clusters that scale down to zero. There you likely want to keep the scheduler always running, but if the cluster hangs e.g. while 100 CPU workers are up because a single GPU worker failed to start, you want to tear it down quickly.

However, I agree that None is generally a more desirable default particularly for non-adaptive situations.

Comment (hendrikmakait) [reply distributed/distributed.yaml]: > Well, there's the use case of adaptive clusters that scale down to zero or almost zero. There you likely want to keep the scheduler always running, but if the cluster hangs e.g. while 100 CPU workers are up because a single GPU worker failed to start, you want to tear it down quickly.

Fair point

File: distributed/scheduler.py
Comment (hendrikmakait) [reply distributed/scheduler.py]: > As a rule of thumb it makes sense to me to keep the variables that hold a public setting as public too.

👍, makes sense to me.

**Review by hendrikmakait — APPROVED** [review]
Thanks! LGTM assuming CI ends up green(ish).

**Pull request merged**
